### ISSUE TITLE
[Bug]: mo panic for free size -1 possible double free when run sysbench and sysbench mixed tesst

### ISSUE THREAD

**Issue by aressu1985** [matrixorigin/matrixone#6607]
### Is there an existing issue for the same bug?

- [X] I have checked the existing issues.

### Environment

```Markdown
- Version or commit-id (e.g. v0.1.0 or 8b23a93):7e5f2f67096c8c5b4e02d51293b3880a328a8841
- Hardware parameters:
- OS type:
- Others:
```

### Actual Behavior

2022/11/14 21:30:54.309621 +0800 INFO frontend/util.go:495 query trace {"connection_id": 1086, "query": "select c from sbtest1 where id = 3514;", "session_info": "connectionId 1086"}
2022/11/14 21:30:54.312592 +0800 INFO frontend/util.go:495 query trace status {"connection_id": 4294967295, "statement": "start transaction", "status": "success", "session_info": ""}
2022/11/14 21:30:54.312705 +0800 INFO frontend/util.go:495 query trace status {"connection_id": 4294967295, "statement": "commit", "status": "success", "session_info": ""}
2022/11/14 21:30:54.313570 +0800 ERROR mpool/mpool.go:558 error: internal error: free size -1, possible double free
github.com/matrixorigin/matrixone/pkg/common/mpool.(*fixedPool).free
        /data1/sudong/mo-regression/matrixone/pkg/common/mpool/mpool.go:558
github.com/matrixorigin/matrixone/pkg/common/mpool.(*MPool).Free
        /data1/sudong/mo-regression/matrixone/pkg/common/mpool/mpool.go:582
github.com/matrixorigin/matrixone/pkg/vm/engine/tae/stl/containers.(*StdVector[...]).Close
        /data1/sudong/mo-regression/matrixone/pkg/vm/engine/tae/stl/containers/stdvec.go:89
github.com/matrixorigin/matrixone/pkg/vm/engine/tae/containers.(*vecBase[...]).Close
        /data1/sudong/mo-regression/matrixone/pkg/vm/engine/tae/containers/vecbase.go:164
github.com/matrixorigin/matrixone/pkg/vm/engine/tae/containers.(*vector[...]).Close
        /data1/sudong/mo-regression/matrixone/pkg/vm/engine/tae/containers/vector.go:191
github.com/matrixorigin/matrixone/pkg/vm/engine/tae/dataio/blockio.BlockRead
        /data1/sudong/mo-regression/matrixone/pkg/vm/engine/tae/dataio/blockio/read.go:82
github.com/matrixorigin/matrixone/pkg/vm/engine/disttae.(*blockReader).Read
        /data1/sudong/mo-regression/matrixone/pkg/vm/engine/disttae/reader.go:81
github.com/matrixorigin/matrixone/pkg/vm/pipeline.(*Pipeline).Run
        /data1/sudong/mo-regression/matrixone/pkg/vm/pipeline/pipeline.go:63
github.com/matrixorigin/matrixone/pkg/sql/compile.(*Scope).Run
        /data1/sudong/mo-regression/matrixone/pkg/sql/compile/scope.go:55
github.com/matrixorigin/matrixone/pkg/sql/compile.(*Scope).MergeRun.func1
        /data1/sudong/mo-regression/matrixone/pkg/sql/compile/scope.go:73
panic: internal error: free size -1, possible double free

goroutine 34136107 [running]:
github.com/matrixorigin/matrixone/pkg/common/mpool.(*fixedPool).free(0xc00030f760, 0xc000f4aac0)
        /data1/sudong/mo-regression/matrixone/pkg/common/mpool/mpool.go:558 +0x89
github.com/matrixorigin/matrixone/pkg/common/mpool.(*MPool).Free(0x0?, {0xc000f4aad0, 0x0?, 0x6500000000?})
        /data1/sudong/mo-regression/matrixone/pkg/common/mpool/mpool.go:582 +0x9f
github.com/matrixorigin/matrixone/pkg/vm/engine/tae/stl/containers.(*StdVector[...]).Close(0xc202a46420?)
        /data1/sudong/mo-regression/matrixone/pkg/vm/engine/tae/stl/containers/stdvec.go:89 +0x39
github.com/matrixorigin/matrixone/pkg/vm/engine/tae/containers.(*vecBase[...]).Close(0xc1376e9018?)
        /data1/sudong/mo-regression/matrixone/pkg/vm/engine/tae/containers/vecbase.go:164 +0x3d
github.com/matrixorigin/matrixone/pkg/vm/engine/tae/containers.(*vector[...]).Close(0xc0b8066000?)
        /data1/sudong/mo-regression/matrixone/pkg/vm/engine/tae/containers/vector.go:191 +0x22
github.com/matrixorigin/matrixone/pkg/vm/engine/tae/dataio/blockio.BlockRead({0x29f8198, 0xc18a3db8f0}, 0x3e3bca0?, {0xc285625f00, 0x2, 0x2}, {0xc18e437e78, 0x2, 0x2}, {0xc05840d110, ...}, ...)
        /data1/sudong/mo-regression/matrixone/pkg/vm/engine/tae/dataio/blockio/read.go:82 +0x4cc
github.com/matrixorigin/matrixone/pkg/vm/engine/disttae.(*blockReader).Read(0xc18b802270, {0xc285625f00?, 0x2?, 0x2?}, 0x3?, 0xc154df1f60?)
        /data1/sudong/mo-regression/matrixone/pkg/vm/engine/disttae/reader.go:81 +0x393
github.com/matrixorigin/matrixone/pkg/vm/pipeline.(*Pipeline).Run(0xc134571f28, {0x29d25a0, 0xc18b802270}, 0xc061826300)
        /data1/sudong/mo-regression/matrixone/pkg/vm/pipeline/pipeline.go:63 +0x145
github.com/matrixorigin/matrixone/pkg/sql/compile.(*Scope).Run(0xc1f8c86500?, 0x29f8128?)
        /data1/sudong/mo-regression/matrixone/pkg/sql/compile/scope.go:55 +0xc8
github.com/matrixorigin/matrixone/pkg/sql/compile.(*Scope).MergeRun.func1(0x3e6c3c0?)
        /data1/sudong/mo-regression/matrixone/pkg/sql/compile/scope.go:73 +0x75
created by github.com/matrixorigin/matrixone/pkg/sql/compile.(*Scope).MergeRun
        /data1/sudong/mo-regression/matrixone/pkg/sql/compile/scope.go:68 +0x18d

### Steps to Reproduce

```Markdown
run sysbench mixed cases and tpcc 10warehouse-10terminals about 3 hours

run mo with etc/launch-tae-CN-tae-DN/launch.toml

how to run sysbench insert test:
1.git clone https://github.com/aressu1985/mo-load.git
2.modify mo addr in mo.yml
3.create database
create database sbtest;
4. git clone https://github.com/aressu1985/mo-sysbench.git
5. install sysbench (eg. yun install sysbench on centos)
6.generate data
cd mo-sysbench & sysbench --mysql-host=127.0.0.1 --mysql-port=6001 --mysql-user=dump --mysql-password=111 oltp_common.lua --tables=10 --table_size=10000 --threads=1 --time=30 --report-interval=10 --create_secondary=off --auto_inc=off prepare

7.run test
./start.sh -c cases/sysbench/mixed_10_10000/

how to run TPCC：
1、https://github.com/aressu1985/mo-tpcc
   启动shell

2、启动MO：
  create database tpcc;
  use tpcc;
  //执行 mo-tpcc/sql/tableCreates_pk.sql 的建表语句

  应该也可以用  source /path/to/mo-tpcc/sql/tableCreates_pk.sql; 建库表

3、在 mo-tpcc下执行
   mkdir data
   ./runLoader.sh props.mo warehouse 1 filelocation data/

   //记住这里的  1 ， 
   //这样会在 data 目录生成要导入的csv文件

4、导入初始化数据
load data infile '/path/to/mo-tpcc/data/config.csv' into table bmsql_config;
load data infile '/path/to/mo-tpcc/data/cust-hist.csv' into table bmsql_history;
load data infile '/path/to/mo-tpcc/data/customer.csv' into table bmsql_customer;
load data infile '/path/to/mo-tpcc/data/district.csv' into table bmsql_district;
load data infile '/path/to/mo-tpcc/data/item.csv' into table bmsql_item;
load data infile '/path/to/mo-tpcc/data/new-order.csv' into table bmsql_new_order;
load data infile '/path/to/mo-tpcc/data/order-line.csv' into table bmsql_order_line;
load data infile '/path/to/mo-tpcc/data/order.csv' into table bmsql_oorder;
load data infile '/path/to/mo-tpcc/data/stock.csv' into table bmsql_stock;
load data infile '/path/to/mo-tpcc/data/warehouse.csv' into table bmsql_warehouse;

5、修改 props.mo
  里面的 warehouses要等于3种的那个值，比如这里的10.
  terminals 必须是 warehouses 的整数倍。且小于10倍，比如50
  runMins 是执行时长，比如1就是一直跑1分钟
6、执行./runBenchmark.sh props.mo
```

**Comment by XuPeng-SH** [comment]
I think it is a mpool bug. It can be solid reproduced by the below code.
```go
package main

import (
    "sync"

    "github.com/matrixorigin/matrixone/pkg/common/mpool"
)

func main() {
    pool, err := mpool.NewMPool("default", 0, mpool.Large)
    if err != nil {
        panic(err)
    }
    var wg sync.WaitGroup
    run := func() {
        defer wg.Done()
        for i := 0; i < 1000; i++ {
            buf, err := pool.Alloc(10)
            if err != nil {
                panic(err)
            }
            pool.Free(buf)
        }
    }
    for i := 0; i < 3000; i++ {
        wg.Add(1)
        go run()
    }
    wg.Wait()
}
```

**Comment by m-schen** [comment]
It seems that Mpool has a bug. TAE teem make an UT can repro the bug just using mpool.Alloc and mpool.Free with high concurrency.

**Comment by XuPeng-SH** [comment]
#6142 same issue

**Comment by LeftHandCold** [comment]
WARNING: DATA RACE
Write at 0x00c000be4a00 by goroutine 383:
  runtime.slicecopy()
      /usr/local/go/src/runtime/slice.go:307 +0x0
  github.com/matrixorigin/matrixone/pkg/common/mpool.(*fixedPool).alloc()
      /Users/shenjiangwei/Work/code/view/matrixone/pkg/common/mpool/mpool.go:499 +0x128
  github.com/matrixorigin/matrixone/pkg/common/mpool.(*MPool).Alloc()
      /Users/shenjiangwei/Work/code/view/matrixone/pkg/common/mpool/mpool.go:516 +0x15c
  github.com/matrixorigin/matrixone/pkg/common/mpool.TestMP.func1()
      /Users/shenjiangwei/Work/code/view/matrixone/pkg/common/mpool/mpool_test.go:177 +0xb0

Previous write at 0x00c000be4a00 by goroutine 77:
  runtime.slicecopy()
      /usr/local/go/src/runtime/slice.go:307 +0x0
  github.com/matrixorigin/matrixone/pkg/common/mpool.(*fixedPool).alloc()
      /Users/shenjiangwei/Work/code/view/matrixone/pkg/common/mpool/mpool.go:499 +0x128
  github.com/matrixorigin/matrixone/pkg/common/mpool.(*MPool).Alloc()
      /Users/shenjiangwei/Work/code/view/matrixone/pkg/common/mpool/mpool.go:516 +0x15c
  github.com/matrixorigin/matrixone/pkg/common/mpool.TestMP.func1()
      /Users/shenjiangwei/Work/code/view/matrixone/pkg/common/mpool/mpool_test.go:177 +0xb0

**Comment by LeftHandCold** [comment]
I changed the type of "allocSz" and added UT but the data race still occurred
```go
type memHdr struct {
	poolId       int64
	allocSz      atomic.Int32
	fixedPoolIdx int8
	guard        [3]uint8
}

func TestMP(t *testing.T) {
	pool, err := NewMPool("default", 0, Large)
	if err != nil {
		panic(err)
	}
	var wg sync.WaitGroup
	run := func() {
		defer wg.Done()
		for i := 0; i < 1000; i++ {
			buf, err := pool.Alloc(10)
			if err != nil {
				panic(err)
			}
			pool.Free(buf)
		}
	}
	for i := 0; i < 800; i++ {
		wg.Add(1)
		go run()
	}
	wg.Wait()

}
```

**Comment by yingfeng** [comment]
freelist could not guarantee concurrent getting head safely(two goroutines might get same head).
```
pos := fl.head.Load()
next := fl.next(pos)
ok := fl.head.CompareAndSwap(pos, next)
```
Goroutine 1:
```
pos = 1
next = 2
head = 2
```
Goroutine 2:
```
pos = 1
next = 2
head = 2
```
As a result the two goroutines have the same head memory causing race. Another race is the alloc counter

**Comment by nnsgmsone** [comment]
The problem with mpool is not actually caused by a lock-free queue, but essentially for the following reasons. 
Suppose the queue length is 2. At the beginning there exist two elements [x, y], then both x, y are allocated, then the queue is empty i.e. [], at this time x is released and inside the queue is [x], at this time x is allocated again, then x is released again and inside the queue is [x, x]. mpool's core problem is the wrong assumption of lock-free, the assumption should be that mpool is a collection and all elements are distinct. The current lock-free setting of mpool will break this assumption and lead to double free.

**Comment by fengttt** [comment]
Hmm, I see.   Let me take another look.

**Comment by nnsgmsone** [comment]
ok

**Comment by aressu1985** [comment]
test it tomorrow,

**Comment by aressu1985** [comment]
it is in testing

**Comment by aressu1985** [comment]
fixed:622d40ed44fb8b745048f22629e4553da127731e
